Instructor: Session: Disable submit button after session is closed #8368

sukanta-27 opened the pull request:
Fixes #8368 

**Outline of Solution:** Modified the `isAllowedToSubmit` variable so that the `submit` button does not get enabled until the session is open. Now the `submit` button is enabled as long as the session is open. In other situations like `Awaiting` and `Closed` the button in disabled.

**UI Before:**
![screenshot 313](https://user-images.githubusercontent.com/10681755/35513670-becb239a-04fb-11e8-8428-7058325bfc4f.png)

**UI After:**
![screenshot 314](https://user-images.githubusercontent.com/10681755/35513689-c965cc88-04fb-11e8-9e80-c18c56350bcf.png)
<br>
<s>**Not yet Ready for review**</s>

sukanta-27 commented:
Ready for a review

sukanta-27 commented:
@wkurniawan07 As I have worked with you in my previous PRs, will you mind reviewing this PR?

wkurniawan07 commented on src/main/java/teammates/ui/template/InstructorFeedbackSessionActions.java:
```diff
@@ -58,7 +58,7 @@ public InstructorFeedbackSessionActions(PageData data, FeedbackSessionAttributes
         }
 
-        this.isAllowedToSubmit = (session.isVisible() || session.isPrivateSession()) && shouldEnableSubmitLink;
+        this.isAllowedToSubmit = session.isOpened() && shouldEnableSubmitLink;
```
There is a valid use case for allowing to submit before open time: to preview the questions. The condition used should be "(is visible or is private session) *and is not closed*".

sukanta-27 replied:
>The condition used should be "(is visible or is private session) and is not closed"

I implemented this at first, but as the use case you specified slipped my mind, I changed it to what it is right now, thus `this.isAllowedToSubmit = session.isOpened() && shouldEnableSubmitLink;`  . 

Making the changes.

sukanta-27 commented:
@wkurniawan07 The test that is failing in the Travis CI is `InstructorFeedbackSessionsPageUiTest.testAddDeleteActions`, which can be solved using GodMode, but the problem is that, this test is passing on my system. I have tried cleaning and building the project again, restarting the server but the test still passes. I have also tried running it multiple times, but is passes every time (except occasional `TimeOut Exceptions`.  How should I solve this?  

The error is : 
```
1016c1016
<               <a class="btn btn-default btn-xs btn-tm-actions session-submit-for-test margin-bottom-7px" disabled="" href="/page/instructorFeedbackSubmissionEditPage?courseid=CFeedbackUiT.CS2104&fsname=private+session+of+characters1234567+%23&user=CFeedbackUiT.instructor">
---
>               <a class="btn btn-default btn-xs btn-tm-actions session-submit-for-test margin-bottom-7px" href="/page/instructorFeedbackSubmissionEditPage?courseid=CFeedbackUiT.CS2104&fsname=private+session+of+characters1234567+%23&user=CFeedbackUiT.instructor">
	at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:101)
	at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:108)
	at teammates.test.driver.HtmlHelper.assertSameHtml(HtmlHelper.java:95)
	at teammates.test.driver.HtmlHelper.assertSameHtml(HtmlHelper.java:63)
	at teammates.test.pageobjects.AppPage.verifyHtmlPart(AppPage.java:865)
	at teammates.test.pageobjects.AppPage.verifyHtmlMainContent(AppPage.java:917)
	at teammates.test.cases.browsertests.InstructorFeedbackSessionsPageUiTest.testDeleteAction(InstructorFeedbackSessionsPageUiTest.java:569)
	at teammates.test.cases.browsertests.InstructorFeedbackSessionsPageUiTest.testAddDeleteActions(InstructorFeedbackSessionsPageUiTest.java:99)
	...
Failed suite [Failed suite [Failed suite [ci-tests]]] > parallel-ui-tests(failed)(failed)(failed) > teammates.test.cases.browsertests.InstructorFeedbackSessionsPageUiTest.testAddDeleteActions FAILED
    java.lang.AssertionError at InstructorFeedbackSessionsPageUiTest.java:569
```

wkurniawan07 commented:
@sukanta-27 it looks like you're making the same mistake as #8381?

sukanta-27 commented:
@wkurniawan07 But I updated the js files before closing #8381, and also checked for misalignment  in this branch (that was present before updating the js) and it was fixed. I also cleaned the project after that, but this issue persists.

sukanta-27 commented:
@wkurniawan07 There is only one case where this difference in output is occuring, apparently in `Private session`s that does not have `End Time` specified, the `Submit` button remains inactive in the local machine, But not in Travis CI tests. And I am not sure why that is happening. 

I ran `npm run build` again to check if this still persists.

sukanta-27 commented:
@wkurniawan07 
There are two cases where unexpected outcome is occuring
- In case of `Awaiting` sessioins,  `isVisible()` is returning `false`, thus disabling the submit button
- In case of `Private` sessions with no specified `End Time`, `isClosed()` is returning `true` and is disabling the submit button. 

I checked it in the `master` branch, where all `Awaiting` sessions are disabled. but when I checked the documentation of `isVisible()`, it clearly states that 
>Returns true if the session is visible; false if not. Does not care if the session has started or not.

I don't know why this behavior is occuring, I ran `npm run build` after updating the master branch, cleaned the project, ( also Started working in a new branch to check if this branch is corrupted or not) but this behavior persists.

sukanta-27 commented:
@wkurniawan07 @whipermr5 
**Help Needed:** I can't figure out why the test `InstructorFeedbackSessionsPageUiTest.testAddDeleteActions` is failing in the Travis CI but passing locally everytime. I have tried running the test locally multiple times over the last few days, ran `npm run build`, cleaned the project and rebuilt it, but with no result whatsoever. A little hint or suggestion on where I am making a mistake will be very helpful.

sukanta-27 commented:
Ready for a review.

**Outline of solution:** The condition for `isAllowedToSubmit` which I initially applied : 
`this.isAllowedToSubmit = (session.isVisible() || session.isPrivateSession()) && !session.isClosed() && shouldEnableSubmitLink;`  raised one problem and that was, the `Private Sessions` without a specified `End Time` had the `submit` button disabled even when the session was open, as `isClosed()` method checks if the current time is after the session end time, and without the end time being specified, the method returns `true`. So instead, I enabled the submit button for all the `Private session` in the current solution. 

Although we can modify the condition so that the submit button gets enabled when a Private session (with no specified `endTime` given) is open, but that (I thought) would unnecessarily complicate the condition without adding much value. (As, to my understanding, a private session with no specified end time is a rare case)

wkurniawan07 commented:
Let's wait for how #8406 develops then.

sukanta-27 commented:
@wkurniawan07 Sure.

sukanta-27 commented:
@wkurniawan07 
As the issue with `isClosed()` method which was the main focus of #8406 has been fixed by @whipermr5, can we continue this now?
